source/mesh_deformation/interface.cc

```diff
@@ -603,100 +603,295 @@ namespace aspect
+                                  additional_data);
+        mf_mesh_matrix.initialize(system_mf_storage);
+
+        // correct rhs:
+        dealii::LinearAlgebra::distributed::Vector<double> u0, corrected_rhs, mf_solution;
+        mf_mesh_matrix.initialize_dof_vector(u0);
+        mf_mesh_matrix.initialize_dof_vector(corrected_rhs);
+        mf_mesh_matrix.initialize_dof_vector(mf_solution);
+        u0 = 0;
+        mesh_velocity_constraints.distribute(u0);
+        u0.update_ghost_values();
+
+        corrected_rhs = 0;
+
+        FEEvaluation<dim, mesh_deformation_fe_degree, mesh_deformation_fe_degree + 1, dim, double> phi(*mf_mesh_matrix.get_matrix_free());
+        for (unsigned int cell = 0;
+             cell < mf_mesh_matrix.get_matrix_free()->n_cell_batches();
+             ++cell)
+          {
+            phi.reinit(cell);
+            phi.read_dof_values_plain(u0);
+            phi.evaluate(EvaluationFlags::gradients);
+            for (unsigned int q = 0; q < phi.n_q_points; ++q)
+              {
+                phi.submit_gradient(-1.0 * phi.get_gradient(q), q);
+              }
+            phi.integrate(EvaluationFlags::gradients);
+            phi.distribute_local_to_global(corrected_rhs);
           }
+        corrected_rhs.compress(VectorOperation::add);
 
-      rhs.compress (VectorOperation::add);
-      mesh_matrix.compress (VectorOperation::add);
-
-      // Make the AMG preconditioner
-      std::vector<std::vector<bool>> constant_modes;
-      DoFTools::extract_constant_modes (mesh_deformation_dof_handler,
-                                        ComponentMask(dim, true),
-                                        constant_modes);
-      // TODO: think about keeping object between time steps
-      LinearAlgebra::PreconditionAMG preconditioner_stiffness;
-      LinearAlgebra::PreconditionAMG::AdditionalData Amg_data;
-      Amg_data.constant_modes = constant_modes;
-      Amg_data.elliptic = true;
-      Amg_data.higher_order_elements = false;
-      Amg_data.smoother_sweeps = 2;
-      Amg_data.aggregation_threshold = 0.02;
-      preconditioner_stiffness.initialize(mesh_matrix);
-
-      // we solve with higher accuracy in the initial timestep:
-      const double tolerance
-        = sim.parameters.linear_stokes_solver_tolerance
-          * ((this->simulator_is_past_initialization()) ? 1.0 : 1e-5);
+        // setup GMG, follow step-37:
+        MGLevelObject<SystemMatrixType> mg_matrices;
+        mesh_deformation_dof_handler.distribute_mg_dofs();
+        MGConstrainedDoFs mg_constrained_dofs;
+        mg_constrained_dofs.initialize(mesh_deformation_dof_handler);
+
+        // Not sure if we need to consider other boundary ids here.
+        mg_constrained_dofs.make_zero_boundary_constraints(mesh_deformation_dof_handler,
+                                                           zero_mesh_deformation_boundary_indicators);
+        mg_constrained_dofs.make_zero_boundary_constraints(mesh_deformation_dof_handler,
```

Comment:
I don't think you are handling the constraints correctly.  See make_constraints():
- periodic constraints (ignore for now, but add an ``Assert(false, ExcNotImeplemented);``
- 0 for zero_mesh_deformation_boundary_indicators
- no normal flux for tangential_mesh_deformation_boundary_indicators
- all other ids get their values from the plugins using compute_velocity_constraints_on_boundary

Note that the last kind has inhomogenous constraints.